cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java

```diff
@@ -737,22 +742,50 @@ public QueryHandle createNamespace(Id.Namespace namespace) throws ExploreExcepti
-
-        String database = getHiveDatabase(namespace.getId());
-        // "IF NOT EXISTS" so that this operation is idempotent.
-        String statement = String.format("CREATE DATABASE IF NOT EXISTS %s", database);
-        operationHandle = doExecute(sessionHandle, statement);
-        QueryHandle handle = saveReadOnlyOperation(operationHandle, sessionHandle, sessionConf, statement, database);
-        LOG.info("Creating database {} with handle {}", namespace, handle);
+        QueryHandle handle;
+        if (Strings.isNullOrEmpty(customHiveDatabase)) {
+          // if no custom hive database was provided get the hive database according to cdap format and create it
+          // if one does not exists since cdap is responsible for managing the lifecycle of such databases
+          String database = getHiveDatabase(namespace.getId());
+          // "IF NOT EXISTS" so that this operation is idempotent.
+          String statement = String.format("CREATE DATABASE IF NOT EXISTS %s", database);
+          operationHandle = doExecute(sessionHandle, statement);
+          handle = saveReadOnlyOperation(operationHandle, sessionHandle, sessionConf, statement, database);
+          LOG.info("Creating database {} with handle {}", namespace, handle);
+        } else {
+          // a custom database name was provided so check its existence
+          // there is no way to check if a hive database exists or not other than trying to use it and see whether
+          // it fails or not. So, run a USE databaseName command and see if it throws exception
+          // Other way can be to list all database and check if the database exists or not but we are doing USE to
+          // make sure that user can acutally use the database once we have impersonation.
+          String statement = String.format("USE %s", customHiveDatabase);
+          // if the database does not exists the below line will throw exception from hive
+          operationHandle = doExecute(sessionHandle, statement);
+          // if we didn't got an exception on the line above we know that the database exists
+          handle = saveReadOnlyOperation(operationHandle, sessionHandle, sessionConf, statement,
+                                                     customHiveDatabase);
+          LOG.debug("Custom database {} existence verified with handle {}", customHiveDatabase, handle);
+        }
         return handle;
       } catch (Throwable e) {
         closeInternal(getQueryHandle(sessionConf),
                       new ReadOnlyOperationInfo(sessionHandle, operationHandle, sessionConf, "", ""));
         throw e;
       }
     } catch (HiveSQLException e) {
+      // Earlier we checked if the hive database exists or not for custom mapped namespacce. If it did not exists
+      // then we will get an exception from Hive with error code 10072 which represent database was not found
+      if (e.toTStatus().getErrorCode() == ErrorMsg.DATABASE_NOT_EXISTS.getErrorCode()) {
```

Comment:
if it was some other error, like an invalid database name or a transient error, this will not catch the error. Can we just try catch around the statement instead of catching here and checking error code?